log/sockstatlog/logger.go

```diff
@@ -0,0 +1,135 @@
+	"path/filepath"
+	"time"
+
+	"tailscale.com/logtail/filch"
+	"tailscale.com/net/sockstats"
+)
+
+// pollPeriod specifies how often to poll for socket stats.
+const pollPeriod = time.Second / 10
+
+// Logger logs statistics about network sockets.
+type Logger struct {
+	ticker time.Ticker
+
+	// last is the last set of socket stats we saw.
+	lastStats *sockstats.SockStats
+	lastTime  time.Time
+
+	logbuffer *filch.Filch
+}
+
+// deltaStat represents the bytes transmitted and received during a time period.
+type deltaStat struct {
+	TxBytes uint64 `json:"tx,omitempty"`
+	RxBytes uint64 `json:"rx,omitempty"`
+}
+
+// event represents the socket stats on a specific interface during a time period.
+type event struct {
+	// Time is when the event started as a Unix timestamp in milliseconds.
+	Time int64 `json:"t"`
+
+	// Duration is the duration of this event in milliseconds.
+	Duration int64 `json:"d"`
+
+	// Interface is the network interface that the event occurred on.
+	Interface string `json:"i"`
+
+	// Stats records the stats for each Label during the time period.
+	Stats map[sockstats.Label]deltaStat `json:"s"`
+}
+
+// NewLogger returns a new Logger that will store stats in logdir.
+// On platforms that do not support sockstat logging, a nil Logger will be returned.
+// The returned Logger is not yet running.
+func NewLogger(logdir string) (*Logger, error) {
+	if !loggingEnabled {
+		return nil, nil
+	}
+	filchPrefix := filepath.Join(logdir, "sockstats")
+	filch, err := filch.New(filchPrefix, filch.Options{ReplaceStderr: false})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Logger{
+		ticker:    *time.NewTicker(pollPeriod),
+		logbuffer: filch,
+	}, nil
+}
+
+func (l *Logger) Start() {
+	go l.poll()
+}
+
+// poll fetches the current socket stats at the configured time interval,
+// calculates the delta since the last poll, and logs any non-zero values.
+// This method does not return.
+func (l *Logger) poll() {
+	enc := json.NewEncoder(l.logbuffer)
+	for t := range l.ticker.C {
+		stats := sockstats.Get()
+		if l.lastStats != nil {
+			diffstats, iface := delta(l.lastStats, stats)
+			if len(diffstats) > 0 {
+				entry := event{
+					Time:      l.lastTime.UnixMilli(),
+					Duration:  t.Sub(l.lastTime).Milliseconds(),
+					Interface: iface,
+					Stats:     diffstats,
+				}
+				enc.Encode(entry)
+			}
+		}
+		l.lastTime = t
+		l.lastStats = stats
+	}
+}
+
+func (l *Logger) Shutdown() {
+	l.ticker.Stop()
+	l.logbuffer.Close()
+}
+
+// delta calculates the delta stats between two SockStats snapshots.
+// b is assumed to have occurred after a.
+// Zero values are omitted from the returned map, and an empty map is returned if no bytes were transferred.
+// iface is the first network interface found that bytes were transferred on.
+func delta(a, b *sockstats.SockStats) (stats map[sockstats.Label]deltaStat, iface string) {
+	if a == nil || b == nil {
+		return nil, ""
+	}
+	stats = make(map[sockstats.Label]deltaStat, len(b.Stats))
```

Comment:
Can use `mak.Set` to avoid allocating the map if it ends up being empty.

Reply:
done